### PR TITLE
Stop LDAP Auth from succeeding without password

### PR DESCRIPTION
**Pull request by keegan**
Fixing issue #1724

### CODE REVIEW

**Comment by astorije** [comment]
@eliemichel, @thisisdarshan, @gramakri, @lindskogen, pinging you here as well, sorry for the spam! 🙏

**Comment by xPaw** [comment]
Sounds rather odd, isn't it your LDAP server that doesn't return an error if the password is empty?

Could you also add a test case for this change?

**Comment by dgw** [comment]
It's default behavior for some LDAP servers to allow anonymous binds (`unauthenticated`, in spec terms). The spec also [specifically warns](https://tools.ietf.org/html/rfc4513#section-6.3.1) against clients trusting the LDAP server's bind result without checking if the password is empty:

> Operational experience shows that clients can (and frequently do) misuse the unauthenticated authentication mechanism of the simple Bind method (see Section 5.1.2).  For example, a client program might make a decision to grant access to non-directory information on the basis of successfully completing a Bind operation.  LDAP server implementations may return a success response to an unauthenticated Bind request.  This may erroneously leave the client with the impression that the server has successfully authenticated the identity represented by the distinguished name when in reality, an anonymous authorization state has been established.  Clients that use the results from a simple Bind operation to make authorization decisions should actively detect unauthenticated Bind requests (by verifying that the supplied password is not empty) and react appropriately.

tl;dr this check for empty passwords in LDAP authentication mode should be merged and released ASAP.

**Comment by gramakri** [comment]
On one hand, I am tempted to call this a server configuration issue. As per http://www.openldap.org/doc/admin23/security.html, you can disable anonymous binds on the server side using `disallow bind_anon` in slapd.conf.

On the other hand, most people don't know of this feature and it's very dangerous. The change looks good but please add a comment as to why we are making that check. Maybe a link to https://tools.ietf.org/html/rfc4513#section-5.1.1

**Comment by eliemichel** [comment]
Mmh, I wasn't aware of this being a common issue. My prod server does not allow empty passwords and the unit test code does not have this this problem neither. But I totally agree that if it is a typical error that administrators do, it is part of our responsibility to warn them/protect them.

To add a test for this, you'll simply have to `edit startLdapServer::authorize` in `test/plugins/auth/ldap.js` in order to make the fake test server behave like misconfigured one (i.e. authorizing empty password).

**Comment by astorije** [comment]
Hey hey, does anyone want to help @keegan write a test for this? We can make sure commit authorship is kept if so :)

**Comment by astorije** [comment]
@gramakri, @eliemichel, you've both been very helpful to us and our LDAP support. Would you like to be added to our GitHub organization and official reviewers for the LDAP integration?

You don't have to commit to code anything obviously, it's mainly to help us triage bugs, review/test PRs, etc. 🙏

**Comment by gramakri** [comment]
@astorije Yeah, sure. I can help review LDAP related code. Thanks!

**Comment by astorije** [comment]
@gramakri, awesome! I added you to the org 😊 

Does anyone want to help adding a test for this? 🙏

**Comment by astorije** [comment]
@gramakri, would you be able to add a test for this case? 🙏 🙏 🙏

**Comment by astorije** [comment]
Hey @keegan, we have sticker packs for our contributors now!
If you're interested, go fill the form at https://goo.gl/forms/f5usqAEp5DWoeXC92 🙂
